File: 2023/12/part2.py
import functools
import itertools
import re
from time import perf_counter as pf

# @functools.cache
def count_permutations(symbols, counts, group_loc=0):
    if not symbols:
        return not counts and not group_loc
    results = 0
    possibilities = ['.', '#'] if symbols[0] == '?' else [symbols[0]]
    for p in possibilities:
        if p == '#':
            results += count_permutations(symbols[1:], counts, group_loc + 1)
        else:
            if group_loc > 0:
                if counts and counts[0] == group_loc:
                    results += count_permutations(symbols[1:], counts[1:])
            else:
                results += count_permutations(symbols[1:], counts)
    return results

# def count_permutations(symbol, counts):
#     results = set()
#     possibilities = []
#     for s in symbol:
#         if s == '?':
#             possibilities.append(['#', '.'])
#         else:
#             possibilities.append([s])

#     # print(possibilities)

#     for combo in itertools.product(*possibilities):
#         candidate = ''.join(combo)
#         matches = re.findall(r'#+', candidate)
#         match_lengths = [len(x) for x in matches]
#         if match_lengths == counts:
#             results.add(candidate)

#     return len(results)

with open('condition.txt', 'r') as file:
    start = pf()
    count = 0
    for line in file.read().splitlines():
        springs, schema = line.split()
        schema = tuple([*map(int, schema.split(','))])
        count += count_permutations(springs + '.', schema)

    print(count)
    print(f'{pf() - start:.2f} s')


# class Combo:
#     def __init__(self, unknown_combos, unknowns, schema) -> None:
#         self.curr = 0
#         self.stop = unknown_combos
#         self.bits = unknowns
#         self.schema = schema

#     def __iter__(self):
#         return self

#     def __next__(self):
#         if self.curr >= self.stop:
#             raise StopIteration

#         x = bin(self.curr)[2:].zfill(self.bits)
#         # Om '111' in x:
#         # Avancera fram self.curr så många steg som möjligt för att skippa följande tal med tre konsektuiva ettor.
#         # Annars
#         self.curr += 1

#         return x


# # @functools.cache
# def parse(possibility: list, schema):
#     output = [
#         str(len(springs_in_a_row)) if min(schema) <= str(len(springs_in_a_row)) <= max(schema) else min(schema)
#         for spring_group in possibility
#         for springs_in_a_row in filter(lambda it: it, spring_group.split('0'))
#     ]

#     return output


# @functools.cache
# def gen_combo(unknowns, unknown_combos, i):
#     ret_list = []
#     if i < unknown_combos:
#         ret_list.extend(gen_combo(unknowns, unknown_combos, i+1))

#     ret_list.append(bin(i)[2:].zfill(unknowns))
#     return ret_list
#     # return [bin(combo)[2:].zfill(unknowns) for combo in range(unknown_combos)]

# with open('condition.txt', 'r') as file:
#     count = 0
#     for nbr, line in enumerate(file.read().splitlines()):
#         springs, _, schema = line.partition(' ')
#         schema = schema.split(',')
        
#         ext_springs = ''
#         for i in itertools.repeat(springs, 5):
#             ext_springs += i + '?'
#         springs = ext_springs[:-1]

#         ext_schema = []
#         for i in itertools.repeat(schema, 5):
#             ext_schema.extend(i)
#         schema = ext_schema

#         spring_groups = filter(lambda it: it, springs.split('.'))
#         list_of_combos = []
#         for group in spring_groups:
#             unknowns = len(group.replace('#', ''))
#             unknown_combos = 2 ** unknowns

#             combos = gen_combo(unknowns, unknown_combos, 0)  # [bin(combo)[2:].zfill(unknowns) for combo in range(unknown_combos)]
#             for i, char in enumerate(group):
#                 if char == '#':
#                     combos = list(map(lambda el: el[:i] + '1' + el[i:], combos))

#             list_of_combos.append(combos)

#         list_of_combos = list(map(lambda el: list(set(parse(el, schema))), list_of_combos))

#         possibilities = filter(
#             lambda it: it == schema,
#             itertools.product(*list_of_combos)
#         )

#         for _ in possibilities:
#             count += 1

#         print('31')
#         # count += len(list(possibilities))
#         # for i, possibility in enumerate(possibilities):
#         #     print(i)
#         #     if parse(possibility) == schema:
#         #         count += 1
#         # break
#     print(count)

File: 2023/12/src/part2.rs
use std::{
    fs::File,
    io::Read, fmt::Binary, borrow::BorrowMut,
};
use anyhow::Result;
use itertools::Itertools;

fn parse() {

}

pub fn run() -> Result<u32> {
    let mut buf = String::new();
    File::open("condition.txt")?.read_to_string(&mut buf)?;

    for line in buf.lines() {
        let parsed_line: (&str, &str) = line.split_ascii_whitespace().collect_tuple().unwrap();
        let springs = parsed_line.0;
        let schema: Vec<&str> = parsed_line.1.split(",").collect();

        let mut ext_springs = String::new();
        let mut ext_schema: Vec<&str> = Vec::new();
        for i in 0..5 {
            ext_springs += springs;
            ext_schema.extend_from_slice(&schema);
            if i < 4 {
                ext_springs += "?";
            }
        }

        let spring_groups: Vec<&str> = ext_springs.split(".").filter(|el| *el != "").collect();
        println!("{:?}", spring_groups);
        let mut list_of_combos: Vec<Vec<String>> = Vec::new();
        for group in spring_groups.iter() {
            let unknowns = group.replace("#", "").len() as u32;
            let unknown_combos = 2_u32.pow(unknowns);

            let mut combos: Vec<String> = Vec::new();
            for combo in 0..unknown_combos {
                combos.push(format!("{:0fill$b}", combo, fill = unknowns as usize))
            }
            let mut idx = 0;
            for character in group.chars() {
                if character == '#' {
                    combos.iter_mut().map(|el| el.insert(idx, '1')).collect_vec();
                    idx += 1;
                }
            }
            list_of_combos.push(combos);
            // [bin(combo)[2:].zfill(unknowns) for combo in range(unknown_combos)]
        }

        println!("{:?}", list_of_combos);
        let mut output: Vec<Vec<Vec<usize>>> = Vec::new();
        for possibility in list_of_combos{
            let mut output2: Vec<Vec<usize>> = Vec::new();
            for spring_group in possibility {
                let mut output3: Vec<usize> = Vec::new();
                for springs_in_a_row in spring_group.split("0") {
                    if springs_in_a_row == "" {
                        continue;
                    }
                    output3.push(springs_in_a_row.len());
                }
                if output3.len() == 0 {
                    continue;
                }
                output2.push(output3);
            }
            output.push(output2);
        }
        println!("{:?}", output);
        println!("{:?}", schema);
        break;
    }
    Ok(100)
}